### DISCUSSION TITLE
Move windows to desktop x

### DISCUSSION THREAD

**Discussion by ms4ndst** [dalyIsaac/Whim#974] [Q&A] [answered]
Stupid question but I can't figure out how I configure moving the active window on desktop 1 to desktop 2 for example. I found MoveWindowToAdjacentWorkspace but I am no programmer.....

**Comment by dalyIsaac** [comment]
Not a stupid question at all! There are a few ways to achieve this. I'm not entirely which scenario you were referring to, so I'll try cover them all.

## Move a window to an adjacent monitor

There are the commands `whim.core.move_window_previous_monitor` and `whim.core.move_window_next_monitor`. These have default keybindings associated with them - see the [Commands](https://dalyisaac.github.io/Whim/docs/customize/commands.html) page of the docs.

## Move a window to a specific workspaces

There are two approaches to this:

1. Use the [Command Palette](https://dalyisaac.github.io/Whim/docs/plugins/command-palette.html) to access the "Move window to workspace" command. From there, you can select the workspace you want to move the window to.
2. Code something up in your config, which would be more involved.

## Move window to monitor index

I've made a new release of Whim [v0.7.8-alpha](https://github.com/dalyIsaac/Whim/releases/tag/v0.7.8-alpha%2Bf497291e) which adds an API to make it slightly easier to move windows to a specific monitor.

You can add this to your `whim.config.csx` file to create [commands](https://dalyisaac.github.io/Whim/docs/customize/commands.html) to use this new API.

```csharp
// NOTE: The 2nd monitor has index 1 due to 0-based indexing (the first item has index 0).
context.CommandManager.Add("move_window_to_monitor_2", "Move window to monitor 2", () =>
{
	context.Store.Dispatch(new MoveWindowToMonitorIndexTransform(1));
});
context.KeybindManager.SetKeybind(
	"whim.custom.move_window_to_monitor_2",
	new Keybind(KeyModifiers.LWin | KeyModifiers.LControl, VIRTUAL_KEY.VK_2)
);
```

The monitors are ordered by their x-coordinate and then by their y-coordinate. Your primary monitor will have the top-left coordinate be (0, 0). Accordingly, monitors to the left of the primary monitor will have negative x-coordinates, and monitors above the primary monitor will have negative y-coordinates.

This would've been possible from the config using C# without this API, but I thought it'd be worth adding.

**Comment by ms4ndst** [comment]
Hi! Thanks for the quick reply! And thanks for this awesome project.

I actually tried the first two suggestions already but nothing happens when I press win+shift+RIGHT or use the Command Palette.
I am wondering if there is something blocking this combination on my machine or if there is something else wrong.

Regarding option 3 "Move window to monitor index" I get that I add this code to the whim.config.csx but I don't understand where I then store the commands "whim.custom.move_window_to_monitor_2". I tried some other tiling window managers, non as awesome as this, and they all had a separate config file for keybindings.

Ah are the commands stored in the source? Ok yes I understand, it is. So I need to pull the source, edit the Commands file and build it?

**Reply by dalyIsaac** [reply]
> I actually tried the first two suggestions already but nothing happens when I press win+shift+RIGHT or use the Command Palette. 
> I am wondering if there is something blocking this combination on my machine or if there is something else wrong.

This is very strange - I haven't seen this before. Are any of the keybindings working? Do you have any logs in `.whim`?

> Regarding option 3 "Move window to monitor index" I get that I add this code to the whim.config.csx but I don't understand where I then store the commands "whim.custom.move_window_to_monitor_2". I tried some other tiling window managers, non as awesome as this, and they all had a separate config file for keybindings.

You place the code I specified earlier in the main `DoConfig` block in `whim.config.csx`. For example:

```csharp
// ...

/// <summary>
/// This is what's called when Whim is loaded.
/// </summary>
/// <param name="context"></param>
void DoConfig(IContext context)
{
	// ...
	// ...

	// Set up workspaces.
	context.WorkspaceManager.Add("1");
	context.WorkspaceManager.Add("2");
	context.WorkspaceManager.Add("3");
	context.WorkspaceManager.Add("4");

	// Set up layout engines.
	context.WorkspaceManager.CreateLayoutEngines = () =>
		new CreateLeafLayoutEngine[]
		{
			(id) => SliceLayouts.CreateMultiColumnLayout(context, sliceLayoutPlugin, id, 1, 2, 0),
			(id) => SliceLayouts.CreatePrimaryStackLayout(context, sliceLayoutPlugin, id),
			(id) => SliceLayouts.CreateSecondaryPrimaryLayout(context, sliceLayoutPlugin, id),
			(id) => new FocusLayoutEngine(id),
			(id) => new TreeLayoutEngine(context, treeLayoutPlugin, id),
			(id) => new FloatingLayoutEngine(context, id)
		};

	// NEW LINES START HERE
	context.CommandManager.Add("move_window_to_monitor_2", "Move window to monitor 2", () =>
	{
		context.Store.Dispatch(new MoveWindowToMonitorIndexTransform(1));
	});
	context.KeybindManager.SetKeybind("whim.custom.move_window_to_monitor_2", new Keybind(KeyModifiers.LWin | KeyModifiers.LControl, VIRTUAL_KEY.VK_2));
	// NEW LINES END HERE
}

// We return doConfig here so that Whim can call it when it loads.
return DoConfig;
```

It isn't necessary to edit source code or build Whim - Whim's philosophy is that it lets extensive customization via user scripting.

**Comment by ms4ndst** [comment]
Yes, some commands work like switching to specifik desktop or focus window. Nothing in the log.

I have tested to place the code you gave me in the config file and the stop and start whim but nothing happens.
I have made the bind (I think) to LAlt+CTRL+a number so it should be able to switch to desktop 2 right?

`// NEW LINES START HERE
	context.CommandManager.Add("move_window_to_monitor_2", "Move window to monitor 2", () =>
	{
		context.Store.Dispatch(new MoveWindowToMonitorIndexTransform(1));
	});
	context.KeybindManager.SetKeybind("whim.custom.move_window_to_monitor_2", new Keybind(KeyModifiers.LAlt | KeyModifiers.LControl, VIRTUAL_KEY.VK_2));
	// NEW LINES END HERE`

**Reply by urob** [reply]
> it should be able to switch to desktop 2 right?

Just to clarify, when you say desktop, do you mean a physical monitor or a virtual workspace? 

(In Whim, desktop means the former but I know that in some other context it refers to the latter).

**Comment by ms4ndst** [comment]
Ah ok yes I mean virtual desktop. I use Hyperland on my personal Linux box and there you can move a window from a virtual desktop to another with keybindings. That was what I was after :/

**Reply by urob** [reply]
> Ah ok yes I mean virtual desktop. I use Hyperland on my personal Linux box and there you can move a window from a virtual desktop to another with keybindings. That was what I was after :/

Okay, the command that you used would switch between monitors and isn't expected to do anything on single monitor setups.

**Comment by ms4ndst** [comment]
OK so is there a way to accomplish what I am after with Whim or should I add this as a request?

**Reply by urob** [reply]
There's `whim.command_palette.move_window_to_workspace` from the command palette plugin

**Comment by ms4ndst** [comment]
Ok that works great. Then my next question, is it possible to make a keybinding for this, say LAlt+LCTRL+number?

**Reply by urob** [reply]
It should be possible but afaik requires creating [custom commands](https://dalyisaac.github.io/Whim/docs/customize/commands.html#custom-commands) atm. I think: [`Butler.MoveWindowToWorkspace`](https://dalyisaac.github.io/Whim/api/Whim.IButlerChores.html#Whim_IButlerChores_MoveWindowToWorkspace_Whim_IWorkspace_Whim_IWindow_) should get you there.

**Reply by ms4ndst** [reply]
Cool that there is a solution but sadly I have no idea how to implement this :( 
Thanks anyway.

**Reply by urob** [reply]
@dalyIsaac Maybe a candidate for https://dalyisaac.github.io/Whim/docs/customize/snippets.html?

**Reply by urob** [reply]
Related: https://github.com/dalyIsaac/Whim/issues/670#issuecomment-1843938046

Allowing commands with arguments would be especially nice for cases like this. 

```cs
context.KeybindManager.SetKeybind("whim.move_window_to_workspace_by_idx" 2, new Keybind(mod, VIRTUAL_KEY.VK_2))
```

**Reply by dalyIsaac** [reply, marked as answer]
Good idea @urob.


@ms4ndst, I've added a snippet to the docs to move a window to a workspace - see https://dalyisaac.github.io/Whim/docs/customize/snippets.html#move-a-window-to-a-specific-workspace. Hopefully this helps. 

The snippet requires the latest release ([`v0.7.9-alpha+895c46db`](https://github.com/dalyIsaac/Whim/releases/tag/v0.7.9-alpha%2B895c46db)) as it relies on a change I just made.

**Reply by ms4ndst** [reply]
@urob do u know how to use this new snippet? I understand the jest of it but not how to implement it in my config. Say I wanted to use LeftShift+CTRL+#ofworkspace to move active window.

**Reply by urob** [reply]
https://dalyisaac.github.io/Whim/docs/customize/keybinds.html#creating-keybinds

The snippet defines a command, `"move_window_to_browser_workspace"`, which you can now bind like any other command.

**Reply by eric-driussi** [reply]
All the links here appear to be broken :(
